### PR TITLE
Receive Member Mail

### PR DESCRIPTION
**Pull request by michaelkain**
- local testing report
	- no mail server on my machine, could only test with log output of ms users
	- to better be tested on qualif
	- 1. test: New member added to study, first member, and flag set: receive member mail true
		- save study has worked, no error
		- Email content checked, contains newly added ONE member
	- 2. test: Add multiple members
		- 2 and 5 added, worked
	- 3. test: Remove the user, with the flag receive member mail true, no member to receive mail
		- now OK, as fixed
	- 4. test: two members to receive emails, with 4 new members added
		- one of the two members already exists before
		- could not test locally as the exception is thrown before sending the second email
	- 5. test: Remove only the user with receive member mail flag from study_user: OK
	- 6. test: All members removed, member list cleaned: OK
	- 7. test: Email contents copied and as .html checked in browser, multiple times: looks good now: OK
	- 8. test: Link checked to manage members of study, works: OK

### CODE REVIEW

**Comment by michaelkain** [comment]
Hi Jean-Come, could you maybe test this PR please for the next release too?

**Comment by jcomedouteau** [comment]
Yes !

File: shanoir-ng-studies/src/main/java/org/shanoir/ng/messaging/StudyUserUpdateBroadcastService.java [hunk 1]
```diff
@@ -34,14 +35,11 @@
 	@Autowired
 	private RabbitTemplate rabbitTemplate;
 	
-	
 	public void broadcast(Iterable<StudyUserCommand> commands) throws MicroServiceCommunicationException {
 		try {
 			String str = new ObjectMapper().writeValueAsString(commands);
-			rabbitTemplate.convertAndSend("study-user-exchange", "study-user", str);
-			
+			rabbitTemplate.convertAndSend(RabbitMQConfiguration.STUDY_USER_EXCHANGE, "study-user", str);
```
Comment (jcomedouteau) [inline shanoir-ng-studies/src/main/java/org/shanoir/ng/messaging/StudyUserUpdateBroadcastService.java:41]: Is  "study-user" a queue ? Maybe we can define this constant in RabbitMQConfiguration too ?

File: shanoir-ng-users/src/main/java/org/shanoir/ng/email/EmailServiceImpl.java [hunk 2]
```diff
@@ -417,20 +425,38 @@ public void notifyStudyManagerDataImported(DatasetImportEmail generatedMail) {
...
+				messageHelper.setSubject("[Shanoir] Member(s) added to " + email.getStudyName());
+				final Map<String, Object> variables = new HashMap<>();
+				variables.put(FIRSTNAME, studyAdmin.getFirstName());
+				variables.put(LASTNAME, studyAdmin.getLastName());
+				variables.put(EMAIL, user.getEmail());
+				variables.put(STUDY_NAME, email.getStudyName());
+				variables.put(STUDY_USERS, newStudyUsers);
+				variables.put(SERVER_ADDRESS, shanoirServerAddress + "study/edit/" + email.getStudyId());
+				final String content = build("notifyStudyAdminStudyUsersAdded", variables);
+				LOG.error(content);
```
Comment (jcomedouteau) [inline shanoir-ng-users/src/main/java/org/shanoir/ng/email/EmailServiceImpl.java:453]: - ` LOG.error(content);`
should not be here (but also upper in my code..)

File: db-changes/datasets/28102021_alter_study_user.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+ALTER TABLE study_user RENAME COLUMN receive_anonymization_report TO receive_study_user_report;
```
Comment (jcomedouteau) [inline db-changes/datasets/28102021_alter_study_user.sql:1]: On qualif (maybe mysql is not up to date ? https://stackoverflow.com/questions/30290880/rename-a-column-in-mysql)
ALTER TABLE study_user RENAME COLUMN receive_anonymization_report TO receive_study_user_report; 
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'COLUMN receive_anonymization_report TO receive_study_user_report' at line 1

Comment (jcomedouteau) [reply db-changes/datasets/28102021_alter_study_user.sql:1]: Could be

`ALTER TABLE study_user CHANGE receive_anonymization_report receive_study_user_report bit(1);`

File: db-changes/studies/28102021_alter_study_user.sql [hunk 4]
```diff
@@ -0,0 +1 @@
+ALTER TABLE study_user RENAME COLUMN receive_anonymization_report TO receive_study_user_report;
```
Comment (jcomedouteau) [inline db-changes/studies/28102021_alter_study_user.sql:1]: Also, I think this needs to be done on import too.

File: db-changes/datasets/28102021_alter_study_user.sql (hunk 3 above)
Comment (jcomedouteau) [inline db-changes/datasets/28102021_alter_study_user.sql:1]: Another thing, this file is not up to date with the new database update from anthony. So when I deploy on qualif, it does not apply automatically

File: shanoir-ng-studies/src/main/java/org/shanoir/ng/messaging/StudyUserUpdateBroadcastService.java (hunk 1 above)
Comment (jcomedouteau) [reply shanoir-ng-studies/src/main/java/org/shanoir/ng/messaging/StudyUserUpdateBroadcastService.java:41]: STUDY_USER_QUEUE

Comment (michaelkain) [reply shanoir-ng-studies/src/main/java/org/shanoir/ng/messaging/StudyUserUpdateBroadcastService.java:41]: done, part of my next commit

File: shanoir-ng-users/src/main/java/org/shanoir/ng/email/EmailServiceImpl.java (hunk 2 above)
Comment (michaelkain) [reply shanoir-ng-users/src/main/java/org/shanoir/ng/email/EmailServiceImpl.java:453]: done, changed to info

File: db-changes/datasets/28102021_alter_study_user.sql (hunk 3 above)
Comment (michaelkain) [reply db-changes/datasets/28102021_alter_study_user.sql:1]: done

File: db-changes/studies/28102021_alter_study_user.sql (hunk 4 above)
Comment (michaelkain) [reply db-changes/studies/28102021_alter_study_user.sql:1]: done, added to import too

File: db-changes/datasets/28102021_alter_study_user.sql (hunk 3 above)
Comment (michaelkain) [reply db-changes/datasets/28102021_alter_study_user.sql:1]: moved everything into /docker-compose/database/db-changes

**Comment by michaelkain** [comment]
Hi Jean-Come, thank you! I let you please close the conversations above, if you consider them as fixed. Please let me know if you still use the qualif, that I can test the auto-apply of the db scripts within my PR?

**Comment by jcomedouteau** [comment]
I'm done with the qualif, you can use it !

**Comment by michaelkain** [comment]
Great, thx, I will do that.

**Comment by michaelkain** [comment]
Hi Jean-Come, I will check if the qualif is fully deployed. Would it be okay for you to test my new feature? I have already tested it on my machine, and it should work quite fine. Have you tested on your machine already? if not, would you mind testing on qualif to approve the PR?

**Comment by jcomedouteau** [comment]
I'll test on qualif !

**Comment by jcomedouteau** [comment]
1) When I create a new study with members, there is no mail -> it feels normal but I was wondering

2) When I removed a user, I had this one:
```
<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml">
<head></head>
<body>
    Dear <span>Michael</span> <span>Kain</span>,<br/>
    <br/>
    one or more new members have been added to study <span>HeyNewStudy</span> by <span>jean-come.douteau@univ-grenoble-alpes.fr</span>:<br/>
    <ul>
    	
    </ul>
    Please be aware, that it is your responsibility to manage the members of study <span>HeyNewStudy</span>.<br/>
    Please connect to <a href="https://shanoir-qualif.irisa.fr/shanoir-ng/study/edit/224">Shanoir</a> to manage the members.<br/>
    <br/>
    With kind regards,<br/>
    Shanoir administrator
</body>
</html>
```

Maybe we should check the list emptiness ? Or event send a mail when a user is removed ?

3) One interesting comportment is when you add a new user AND you check the "send user update email" -> Then the newly added user will eventually receive the mail for its addition (it's not wrong but can be discussed).

4) A last thing, maybe we could try not to send mail to a user if he made the changes himself

>     Dear <span>Jean-Come</span> <span>Douteau</span>,<br/>
>     <br/>
>     one or more new members have been added to study <span>HeyNewStudy</span> by <span>jean-come.douteau@univ-grenoble-alpes.fr</span>:<br/>

But this is maybe not the best idea. We may send a mail everytime ?

**Comment by michaelkain** [comment]
Great! As all 4 points discussed, it should be fine now for approval?

**Review by jcomedouteau — APPROVED** [review]

**Pull request merged**